beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -1849,13 +1861,123 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         Ok(())
     }
 
+    fn prune_abandoned_forks(
+        &self,
+        old_finalized_block_hash: SignedBeaconBlockHash,
+        new_finalized_block_hash: SignedBeaconBlockHash,
+        new_finalized_slot: Slot,
+    ) -> Result<(), Error> {
+        let old_finalized_slot = self
+            .get_block(&old_finalized_block_hash.into())?
+            .ok_or_else(|| Error::MissingBeaconBlock(old_finalized_block_hash.into()))?
+            .message
+            .slot;
+
+        // Collect hashes from new_finalized_block up to old_finalized_block
+        let newly_finalized_blocks: HashSet<SignedBeaconBlockHash> = HashSet::from_iter(
+            ParentRootBlockIterator::new(&*self.store, new_finalized_block_hash.into())
+                .take_while(|(block_hash, _)| *block_hash != old_finalized_block_hash.into())
+                .map(|(block_hash, _)| block_hash.into()),
+        );
+
+        // We don't know which blocks are shared among abandoned chains, so we buffer and delete
+        // everything in one fell swoop.
+        let mut abandoned_blocks: HashSet<SignedBeaconBlockHash> = HashSet::new();
+        let mut abandoned_states: HashSet<(Slot, BeaconStateHash)> = HashSet::new();
+        let mut abandoned_heads: HashSet<Hash256> = HashSet::new();
+
+        for (head_hash, head_slot) in self.heads() {
+            let mut potentially_abandoned_head: Option<Hash256> = Some(head_hash);
```

Comment:
I think this var is unnecessary because `potentially_abandoned_head.is_some()` iff `!potentially_abandoned_blocks.is_empty()`, so we can just push the head in the branch that checks `!potentially_abandoned_blocks.is_empty()`.

Reply:
And remove it from `abandoned_heads` if we detect a chain that is still live?  What if there are duplicates?  What if it's not in `abandoned_heads`?  I find it less error-prone to simply have a separate variable.

Reply:
The change I'm suggesting would be exactly equivalent to what the current code does, just without the mutable variable, but I'm also happy to leave it as is.

To expand on what I mean by exactly equivalent: `potentially_abandoned_head` is initially `Some(head_hash)` and will only get set to `None` in the branch where `potentially_abandonded_blocks` gets `clear`ed. That means that in the branch where we've checked `!potentially_abandoned_forks.is_empty()`, we can infer that `potentially_abandoned_head` must be `Some(head_hash)`.